materialize-elasticsearch/driver.go

```diff
@@ -28,124 +40,281 @@ type credentials struct {
+		return fmt.Errorf("number_of_replicas cannot be negative")
 	}
+
 	return c.Credentials.Validate()
 }
 
+// toClient initializes a client for connecting to Elasticsearch, starting the network tunnel if
+// configured.
+func (c config) toClient() (*client, error) {
+	endpoint := c.Endpoint
+
+	// If SSH Endpoint is configured, then try to start a tunnel before establishing connections
+	if c.NetworkTunnel != nil && c.NetworkTunnel.SshForwarding != nil && c.NetworkTunnel.SshForwarding.SshEndpoint != "" {
+		u, err := url.Parse(c.Endpoint)
+		if err != nil {
+			return nil, fmt.Errorf("parsing endpoint URL: %w", err)
+		}
+
+		var sshConfig = &networkTunnel.SshConfig{
+			SshEndpoint: c.NetworkTunnel.SshForwarding.SshEndpoint,
+			PrivateKey:  []byte(c.NetworkTunnel.SshForwarding.PrivateKey),
+			ForwardHost: u.Hostname(),
+			ForwardPort: u.Port(),
+			LocalPort:   "9200",
+		}
+		var tunnel = sshConfig.CreateTunnel()
+
+		if err := tunnel.Start(); err != nil {
+			return nil, fmt.Errorf("error starting network tunnel: %w", err)
+		}
+
+		// If SSH Tunnel is configured, we are going to create a tunnel from localhost:9200 to
+		// address through the bastion server, so we use the tunnel's address.
+		endpoint = "http://localhost:9200"
+	}
+
+	es, err := elasticsearch.NewClient(
+		elasticsearch.Config{
+			Addresses: []string{endpoint},
+			Username:  c.Credentials.Username,
+			Password:  c.Credentials.Password,
+			APIKey:    c.Credentials.ApiKey,
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("creating client: %w", err)
+	}
+
+	return &client{es: es}, nil
+}
+
 type resource struct {
-	Index         string `json:"index" jsonschema_extras:"x-collection-name=true"`
-	DeltaUpdates  bool   `json:"delta_updates" jsonschema:"default=false"`
-	NumOfShards   int    `json:"number_of_shards,omitempty" jsonschema:"default=1"`
-	NumOfReplicas int    `json:"number_of_replicas,omitempty" jsonschema:"default=0"`
+	Index        string `json:"index" jsonschema_extras:"x-collection-name=true"`
+	DeltaUpdates bool   `json:"delta_updates" jsonschema:"default=false"`
+	Shards       *int   `json:"number_of_shards,omitempty"`
 }
 
 func (r resource) Validate() error {
 	if r.Index == "" {
 		return fmt.Errorf("missing Index")
+	} else if r.Shards != nil && *r.Shards < 1 {
+		return fmt.Errorf("number_of_shards must be greater than 0")
 	}
 
-	if r.NumOfShards <= 0 {
-		return fmt.Errorf("number_of_shards is missing or non-positive")
-	}
 	return nil
 }
 
 // GetFieldDocString implements the jsonschema.customSchemaGetFieldDocString interface.
 func (resource) GetFieldDocString(fieldName string) string {
 	switch fieldName {
 	case "Index":
-		return "Name of the ElasticSearch index to store the materialization results."
+		return "Name of the Elasticsearch index to store the materialization results."
 	case "DeltaUpdates":
 		return "Should updates to this table be done via delta updates. Default is false."
```

Comment:
Not disagreeing that the description is pretty bad, but I am reluctant to spend much time working on the wording for 2 reasons:
- It's the exact same wording as all other SQL materializations use. I could see us making a more sweeping change to the wording of everything at once, but it would be sort of weird to have this one individually be different
- Currently the description doesn't show up at all for boolean fields (there's just a check box), so this is basically non-user-facing text anyway